crates/exec-wasmtime/src/runtime/vfs/listen.rs

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! A file system containing listening networking sockets.
+
+use super::super::net::tls;
+use super::super::WasiResult;
+
+use std::any::Any;
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+use std::ops::Deref;
+use std::path::MAIN_SEPARATOR;
+use std::sync::Arc;
+
+use anyhow::Context;
+use cap_std::net::TcpListener;
+use enarx_config::{FileName, ListenFile};
+use rustls::{Certificate, PrivateKey};
+use wasi_common::file::{FdFlags, FileType};
+use wasi_common::{Error, ErrorExt, WasiDir, WasiFile};
+use wasmtime_vfs_dir::Directory;
+use wasmtime_vfs_ledger::InodeId;
+use wasmtime_vfs_memory::{Data, Inode, Link, Node};
+use wiggle::async_trait;
+use zeroize::Zeroizing;
+
+enum Socket {
+    Tcp(Link<()>),
+    Tls(Link<Arc<rustls::ServerConfig>>),
+}
+
+#[async_trait]
+impl Node for Socket {
+    fn to_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync> {
+        self
+    }
+
+    fn parent(&self) -> Option<Arc<dyn Node>> {
+        match self {
+            Self::Tcp(Link { parent, .. }) | Self::Tls(Link { parent, .. }) => parent.upgrade(),
+        }
+    }
+
+    fn filetype(&self) -> FileType {
+        FileType::SocketStream
+    }
+
+    fn id(&self) -> Arc<InodeId> {
+        match self {
+            Self::Tcp(Link { inode, .. }) => inode.id.clone(),
+            Self::Tls(Link { inode, .. }) => inode.id.clone(),
+        }
+    }
+
+    async fn open_dir(self: Arc<Self>) -> WasiResult<Box<dyn WasiDir>> {
+        Err(Error::not_dir())
+    }
+
+    async fn open_file(
+        self: Arc<Self>,
+        path: &str,
+        dir: bool,
+        read: bool,
+        write: bool,
+        flags: FdFlags,
+    ) -> WasiResult<Box<dyn WasiFile>> {
+        if dir {
+            return Err(Error::not_dir());
+        }
+
+        if !read || !write {
+            return Err(Error::perm()); // FIXME(@npmccallum): errno
+        }
+
+        let addr = path
+            .rsplit_terminator(MAIN_SEPARATOR)
+            .next()
+            .ok_or_else(|| Error::invalid_argument().context("failed to parse file name"))?;
+        let (host, port) = addr
+            .split_once(':')
```

Comment:
Up in enarx-config/src/lib.rs it looks like we support `"::9001"` as an address, so how does this parsing work? I'm wary of doing our own bespoke address parsing here, can we use an existing crate?

Reply:
reworked